src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,385 @@
+                auto col_key = results.get_table()->get_column_key(*prop_name);
+                return [col_key](Mixed value, SharedRealm realm) {
+                    auto link = value.get_link();
+                    auto v = realm->read_group()
+                                 .get_table(link.get_table_key())
+                                 ->get_object(link.get_obj_key())
+                                 .get<StringData>(col_key);
+                    return v.size() > 0 ? v.prefix(1) : "";
+                };
+            }
+            else {
+                return [](Mixed value, SharedRealm) {
+                    auto v = value.get_string();
+                    return v.size() > 0 ? v.prefix(1) : "";
+                };
+            }
+        default:
+            throw std::logic_error("Builtin section algorithm not implemented.");
+    }
+}
+
+struct SectionedResultsNotificationHandler {
+public:
+    SectionedResultsNotificationHandler(SectionedResults& sectioned_results, SectionedResultsNotificatonCallback cb,
+                                        util::Optional<size_t> section_filter = util::none)
+        : m_cb(std::move(cb))
+        , m_sectioned_results(sectioned_results)
+        , m_prev_sections(m_sectioned_results.m_sections)
+        , m_prev_row_to_index_path(m_sectioned_results.m_row_to_index_path)
+        , m_section_filter(section_filter)
+    {
+    }
+
+    void before(CollectionChangeSet const&) {}
+    void after(CollectionChangeSet const& c)
+    {
+        auto convert_indices = [&](const IndexSet& indices,
+                                   const std::vector<std::pair<size_t, size_t>>& rows_to_index_path) {
+            std::map<size_t, IndexSet> ret;
+            for (auto index : indices.as_indexes()) {
+                auto& index_path = rows_to_index_path[index];
+                ret[index_path.first].add(index_path.second);
+            }
+            return ret;
+        };
+
+        m_sectioned_results.calculate_sections_if_required();
+
+        auto converted_insertions = convert_indices(c.insertions, m_sectioned_results.m_row_to_index_path);
+        auto converted_modifications = convert_indices(c.modifications, m_prev_row_to_index_path);
+        auto converted_modifications_new =
+            convert_indices(c.modifications_new, m_sectioned_results.m_row_to_index_path);
+        auto converted_deletions = convert_indices(c.deletions, m_prev_row_to_index_path);
+
+        auto section_changes = calculate_sections_to_insert_and_delete();
+
+        std::map<size_t, IndexSet> modifications_to_keep, modifications_to_keep_new;
+        std::map<size_t, size_t> section_identifiers;
+
+        for (auto& section : m_sectioned_results.m_sections) {
+            section_identifiers[section.hash] = section.index;
+        }
+
+        for (auto [section_old, indexes_old] : converted_modifications) {
+            auto it = m_prev_sections.at(section_old);
```

Comment:
`it` is a generic name for an iterator var, but `at()` doesn't return an iterator. Using `std::vector::at` here is also somewhat dubious; it means that it's possible that `section_old` might be out of bounds and if it is you want this to throw an exception.

Reply:
This is now using an iterator so I think `it` is fine? It will also throw if the element cannot be found.